README
Python Cryptography Toolkit (pycrypto)
======================================

NOTE: This is a TurnKey py2 fork - for use with TKLBAM

This is a collection of both secure hash functions (such as SHA256 and
RIPEMD160), and various encryption algorithms (AES, DES, RSA, ElGamal,
etc.).  The package is structured to make adding new modules easy.
This section is essentially complete, and the software interface will
almost certainly not change in an incompatible way in the future; all
that remains to be done is to fix any bugs that show up.  If you
encounter a bug, please report it in the Launchpad bug tracker at

       https://launchpad.net/products/pycrypto/+bugs

An example usage of the SHA256 module is:
>>> from Crypto.Hash import SHA256
>>> hash = SHA256.new()
>>> hash.update('message')
>>> hash.digest()
'\xabS\n\x13\xe4Y\x14\x98+y\xf9\xb7\xe3\xfb\xa9\x94\xcf\xd1\xf3\xfb"\xf7\x1c\xea\x1a\xfb\xf0+F\x0cm\x1d'

An example usage of an encryption algorithm (AES, in this case) is:

>>> from Crypto.Cipher import AES
>>> obj = AES.new('This is a key456', AES.MODE_ECB)
>>> message = "The answer is no"
>>> ciphertext = obj.encrypt(message)
>>> ciphertext
'o\x1aq_{P+\xd0\x07\xce\x89\xd1=M\x989'
>>> obj2 = AES.new('This is a key456', AES.MODE_ECB)
>>> obj2.decrypt(ciphertext)
'The answer is no'

One possible application of the modules is writing secure
administration tools.  Another application is in writing daemons and
servers.  Clients and servers can encrypt the data being exchanged and
mutually authenticate themselves; daemons can encrypt private data for
added security.  Python also provides a pleasant framework for
prototyping and experimentation with cryptographic algorithms; thanks
to its arbitrary-length integers, public key algorithms are easily
implemented.

As of PyCrypto 2.1.0, PyCrypto provides an easy-to-use random number
generator:

>>> from Crypto import Random
>>> rndfile = Random.new()
>>> rndfile.read(16)
'\xf7.\x838{\x85\xa0\xd3>#}\xc6\xc2jJU'

A stronger version of Python's standard "random" module is also
provided:

>>> from Crypto.Random import random
>>> random.choice(['dogs', 'cats', 'bears'])
'bears'

Caveat: For the random number generator to work correctly, you must
call Random.atfork() in both the parent and child processes after
using os.fork()


Installation
============

PyCrypto is written and tested using Python version 2.1 through 3.2.  Python
1.5.2 is not supported.

The modules are packaged using the Distutils, so you can simply run
"python setup.py build" to build the package, and "python setup.py
install" to install it.

If the setup.py script crashes with a DistutilsPlatformError
complaining that the file /usr/lib/python2.2/config/Makefile doesn't
exist, this means that the files needed for compiling new Python
modules aren't installed on your system.  Red Hat users often run into
this because they don't have the python2-devel RPM installed.  The fix
is to simply install the requisite RPM.  On Debian/Ubuntu, you need the
python-dev package.

To verify that everything is in order, run "python setup.py test".  It
will test all the cryptographic modules, skipping ones that aren't
available.  If the test script reports an error on your machine,
please report the bug using the bug tracker (URL given above).  If
possible, track down the bug and include a patch that fixes it,
provided that you are able to meet the eligibility requirements at
http://www.pycrypto.org/submission-requirements/.

It is possible to test a single sub-package or a single module only, for instance
when you investigate why certain tests fail and don't want to run the whole
suite each time. Use "python setup.py test --module=name", where 'name'
is either a sub-package (Cipher, PublicKey, etc) or a module (Cipher.DES,
PublicKey.RSA, etc).
To further cut test coverage, pass also the option "--skip-slow-tests".

To install the package under the site-packages directory of
your Python installation, run "python setup.py install".

If you have any comments, corrections, or improvements for this
package, please report them to our mailing list, accessible via the
PyCrypto website:

    http://www.pycrypto.org/



debian/changelog
python-crypto (2.6.1-13) unstable; urgency=medium

  [ Sebastian Ramacher ]
  * debian/python3-crypto{,dbg}.install: Avoid brace expansion

  [ Jochen Sprickerhof ]
  * debian/patches: Fix use if "is" with literals (Closes: #944843)

 -- Sebastian Ramacher <sramacher@debian.org>  Sat, 16 Nov 2019 10:46:30 +0100

python-crypto (2.6.1-12) unstable; urgency=medium

  * debian/python3-crypto{,-dbg}.install: Fix installation of Python 3.8
    extensions (Closes: #942718)
  * debian/tests: Fix Python 2 test and remove python-all and python-all-dbg
    tests.

 -- Sebastian Ramacher <sramacher@debian.org>  Tue, 22 Oct 2019 21:21:40 +0200

python-crypto (2.6.1-11) unstable; urgency=medium

  * debian/control:
    - Bump Standards-Version
    - Set RRR: no
  * debian/patches: Replace time.clock with timeit.default_timer (Closes:
    #942665)

 -- Sebastian Ramacher <sramacher@debian.org>  Sun, 20 Oct 2019 15:05:21 +0200

python-crypto (2.6.1-10) unstable; urgency=medium

  [ Kenneth J. Pronovici ]
  * Eliminate dependency on epydoc, which will be removed (Closes: #933618)
    - Update debian/rules to no longer build epydoc documentation
    - Update debian/control to remove Build-Depends: python-epydoc
    - Update debian/control to remove now-empty python-crypto-doc package
    - Update debian/control to remove Suggests: python-crypto-doc
    - Remove debian/python-crypto-doc.doc-base
    - Remove debian/python-crypto-doc.docs

  [ Sebastian Ramacher ]
  * debian/control: Bump Standards-Version.
  * debian/: Bump debhelper compat to 12.

 -- Sebastian Ramacher <sramacher@debian.org>  Sun, 04 Aug 2019 20:56:28 +0200

python-crypto (2.6.1-9) unstable; urgency=medium

  * debian/control:
    - Move to salsa.d.o.
    - Bump Standards-Version.
  * debian/: Bump debhelper compat to 11.
  * debian/tests: Add allow-stderr restriction instead of fidling with
    warnings.
  * debian/patches: Apply backported patch from pycryptodome to Fix ElGamal
    key generation (CVE-2018-6594). (Closes: #889999)
  * debian/copyright: Update copyright years.

 -- Sebastian Ramacher <sramacher@debian.org>  Mon, 16 Apr 2018 01:14:05 +0200

python-crypto (2.6.1-8) unstable; urgency=medium

  * debian/control:
    - Run wrap-and-sort.
    - Bump Standards-Version.
    - Remove Andreas from Uploaders. Thank you for all your work on
      python-crypto! (Closes: #880436)
    - Remove Priority: extra.

 -- Sebastian Ramacher <sramacher@debian.org>  Tue, 07 Nov 2017 21:12:04 +0100

python-crypto (2.6.1-7) unstable; urgency=high

  [ Salvatore Bonaccorso ]
  * Throw exception when IV is used with ECB or CTR (CVE-2013-7459)
    (Closes: #849495)

  [ Sebastian Ramacher ]
  * debian/control:
    - Bump Standards-Version.
    - Update Vcs-Git
  * debian/rules: Remove dh_strip override for automatic dbgsym packages.
  * debian/{control,compat}: Bump debhelper compat to 10.

 -- Sebastian Ramacher <sramacher@debian.org>  Wed, 28 Dec 2016 11:45:21 +0100

python-crypto (2.6.1-6) unstable; urgency=medium

  * debian/control:
    - Bump Standards-Version.
    - Remove XS-Testsuite, no longer needed.
    - Remove obsolete Breaks+Replaces
    - Move to collab-maint.
  * debian/rules: Do not run tests if only building arch all packages. (Closes:
    #806866)

 -- Sebastian Ramacher <sramacher@debian.org>  Sun, 06 Dec 2015 15:03:15 +0100

python-crypto (2.6.1-5) unstable; urgency=medium

  * debian/patches/asn1-decoding.patch: Fix TypeError in ASN1 implementation.
    Thanks to Stefanos Harhalakis for the patch. (Closes: #746180)

 -- Sebastian Ramacher <sramacher@debian.org>  Sat, 10 May 2014 19:27:03 +0200

python-crypto (2.6.1-4) unstable; urgency=medium

  * debian/python3-crypto{,-dbg}.install: Fix wildcards to not include debug
    extensions with arm in the name in python3-crypto. (Closes: #740086)
    - Extensions built for Python 3.2 are no longer handled, so this requires
      python3-all-dev >= 3.3.2-5.
  * Use pybuild buildsystem:
    - debian/control: Add dh-python to Build-Depends.
    - debian/rules:
      + Pass --buildsystem=pybuild to dh.
      + Remove dh_auto_clean, dh_auto_build and dh_auto_install overrides.
      + Adapt dh_auto_test override for pybuild.
  * Build documention in -indep targets:
    - debian/control: Move python-docutils and python-epydoc to
      Build-Depends-Indep.
    - debian/rules: Build documentation in override_dh_auto_build-indep.
  * debian/patches/deprecated-test-methods.patch: Replace deprecated unittest
    methods. Python 3.3's and 3.4's 2to3 generate different files otherwise.

 -- Sebastian Ramacher <sramacher@debian.org>  Tue, 25 Feb 2014 23:16:33 +0100

python-crypto (2.6.1-3) unstable; urgency=low

  * debian/control:
    - Fix some mistakes in Description. (Closes: #729241)
    - Bump Standards-Version to 3.9.5.
  * debian/watch: Only honor stable releases.
  * debian/{watch,upstream-signing-key.pgp}: Check signature of the orig
    tarball.
  * debian/patches/unittest-stream.patch: Fix stream handling in SelfTest.
    Otherwise the actual error won't get printed and an unrelated exception
    will be thrown.

 -- Sebastian Ramacher <sramacher@debian.org>  Fri, 24 Jan 2014 00:20:53 +0100

python-crypto (2.6.1-2) unstable; urgency=low

  * debian/patches/multiprocessing-test.patch: Check if
    multiprocessing.synchronize is working before running tests that use it.
    Fixes build failures on kfreebsd-*.

 -- Sebastian Ramacher <sramacher@debian.org>  Thu, 17 Oct 2013 22:02:39 +0200

python-crypto (2.6.1-1) unstable; urgency=high

  [ Jakub Wilk ]
  * Drop obsolete Conflicts/Replaces with python2.3-crypto and
    python2.4-crypto.

  [ Sebastian Ramacher ]
  * New upstream release.
    - Fixes CVE-2013-1445: PRNG not correctly reseeded in some situations.
  * Set urgency to high since this upload fixes a security issue.
  * debian/python3-crypto.install: Fix lintian warning
    brace-expansion-in-debhelper-config-file.


test/test_arc2_buffer_overflow.py
import unittest

class BufferOverflowTest(unittest.TestCase):
    # Test a buffer overflow found in older versions of PyCrypto

    def setUp(self):
        global ARC2
        from Crypto.Cipher import ARC2

    def runTest(self):
        """ARC2 with keylength > 128"""
        key = "x" * 16384
        mode = ARC2.MODE_ECB
        self.assertRaises(ValueError, ARC2.new, key, mode)

if __name__ == '__main__':
    unittest.main()


__init__.py

"""Python Cryptography Toolkit

A collection of cryptographic modules implementing various algorithms
and protocols.

Subpackages:
Crypto.Cipher             Secret-key encryption algorithms (AES, DES, ARC4)
Crypto.Hash               Hashing algorithms (MD5, SHA, HMAC)
Crypto.Protocol           Cryptographic protocols (Chaffing, all-or-nothing
                          transform).   This package does not contain any
                          network protocols.
Crypto.PublicKey          Public-key encryption and signature algorithms
                          (RSA, DSA)
Crypto.Util               Various useful modules and functions (long-to-string
                          conversion, random number generation, number
                          theoretic functions)
"""

__all__ = ['Cipher', 'Hash', 'Protocol', 'Util']

__version__ = '2.0'
__revision__ = "$Id: __init__.py,v 1.10 2004/08/13 23:44:47 akuchling Exp $"




src/ARC2.c

/*
 *  rc2.c : Source code for the RC2 block cipher
 *
 * Part of the Python Cryptography Toolkit
 *
 * Distribute and use freely; there are no restrictions on further 
 * dissemination and usage except those imposed by the laws of your 
 * country of residence.
 *
 */

#include <string.h>  
#include "Python.h"

#define MODULE_NAME ARC2
#define BLOCK_SIZE 8
#define KEY_SIZE 0

typedef unsigned int U32;
typedef unsigned short U16;
typedef unsigned char U8;

typedef struct 
{
	U16 xkey[64];
} block_state;

static void
block_encrypt(block_state *self, U8 *in, U8 *out)
{
	U16 x76, x54, x32, x10;
	int i;
  
	x76 = (in[7] << 8) + in[6];
	x54 = (in[5] << 8) + in[4];
	x32 = (in[3] << 8) + in[2];
	x10 = (in[1] << 8) + in[0];
  
	for (i = 0; i < 16; i++)
	{
		x10 += (x32 & ~x76) + (x54 & x76) + self->xkey[4*i+0];
		x10 = (x10 << 1) + (x10 >> 15 & 1);
      
		x32 += (x54 & ~x10) + (x76 & x10) + self->xkey[4*i+1];
		x32 = (x32 << 2) + (x32 >> 14 & 3);
      
		x54 += (x76 & ~x32) + (x10 & x32) + self->xkey[4*i+2];
		x54 = (x54 << 3) + (x54 >> 13 & 7);
      
		x76 += (x10 & ~x54) + (x32 & x54) + self->xkey[4*i+3];
		x76 = (x76 << 5) + (x76 >> 11 & 31);
      
		if (i == 4 || i == 10) {
			x10 += self->xkey[x76 & 63];
			x32 += self->xkey[x10 & 63];
			x54 += self->xkey[x32 & 63];
			x76 += self->xkey[x54 & 63];
		}
	}
  
	out[0] = (U8)x10;
	out[1] = (U8)(x10 >> 8);
	out[2] = (U8)x32;
	out[3] = (U8)(x32 >> 8);
	out[4] = (U8)x54;
	out[5] = (U8)(x54 >> 8);
	out[6] = (U8)x76;
	out[7] = (U8)(x76 >> 8);
}


static void
block_decrypt(block_state *self, U8 *in, U8 *out)
{
	U16 x76, x54, x32, x10;
	int i;
  
	x76 = (in[7] << 8) + in[6];
	x54 = (in[5] << 8) + in[4];
	x32 = (in[3] << 8) + in[2];
	x10 = (in[1] << 8) + in[0];
  
	i = 15;
	do {
		x76 &= 65535;
		x76 = (x76 << 11) + (x76 >> 5);
		x76 -= (x10 & ~x54) + (x32 & x54) + self->xkey[4*i+3];
    
		x54 &= 65535;
		x54 = (x54 << 13) + (x54 >> 3);
		x54 -= (x76 & ~x32) + (x10 & x32) + self->xkey[4*i+2];
    
		x32 &= 65535;
		x32 = (x32 << 14) + (x32 >> 2);
		x32 -= (x54 & ~x10) + (x76 & x10) + self->xkey[4*i+1];
    
		x10 &= 65535;
		x10 = (x10 << 15) + (x10 >> 1);
		x10 -= (x32 & ~x76) + (x54 & x76) + self->xkey[4*i+0];
    
		if (i == 5 || i == 11) {
			x76 -= self->xkey[x54 & 63];
			x54 -= self->xkey[x32 & 63];
			x32 -= self->xkey[x10 & 63];
			x10 -= self->xkey[x76 & 63];
		}
	} while (i--);
  
	out[0] = (U8)x10;
	out[1] = (U8)(x10 >> 8);
	out[2] = (U8)x32;
	out[3] = (U8)(x32 >> 8);
	out[4] = (U8)x54;
	out[5] = (U8)(x54 >> 8);
	out[6] = (U8)x76;
	out[7] = (U8)(x76 >> 8);
}


static void 
block_init(block_state *self, U8 *key, int keylength)
{
	U8 x;
	U16 i;
        /* 256-entry permutation table, probably derived somehow from pi */
        static const U8 permute[256] = {
		217,120,249,196, 25,221,181,237, 40,233,253,121, 74,160,216,157,
		198,126, 55,131, 43,118, 83,142, 98, 76,100,136, 68,139,251,162,
		23,154, 89,245,135,179, 79, 19, 97, 69,109,141,  9,129,125, 50,
		189,143, 64,235,134,183,123, 11,240,149, 33, 34, 92,107, 78,130,
		84,214,101,147,206, 96,178, 28,115, 86,192, 20,167,140,241,220,
		18,117,202, 31, 59,190,228,209, 66, 61,212, 48,163, 60,182, 38,
		111,191, 14,218, 70,105,  7, 87, 39,242, 29,155,188,148, 67,  3,
		248, 17,199,246,144,239, 62,231,  6,195,213, 47,200,102, 30,215,
		8,232,234,222,128, 82,238,247,132,170,114,172, 53, 77,106, 42,
		150, 26,210,113, 90, 21, 73,116, 75,159,208, 94,  4, 24,164,236,
		194,224, 65,110, 15, 81,203,204, 36,145,175, 80,161,244,112, 57,
		153,124, 58,133, 35,184,180,122,252,  2, 54, 91, 37, 85,151, 49,
		45, 93,250,152,227,138,146,174,  5,223, 41, 16,103,108,186,201,
		211,  0,230,207,225,158,168, 44, 99, 22,  1, 63, 88,226,137,169,
		13, 56, 52, 27,171, 51,255,176,187, 72, 12, 95,185,177,205, 46,
		197,243,219, 71,229,165,156,119, 10,166, 32,104,254,127,193,173
        };

	/* The "bits" value may be some sort of export control weakening.
	   We'll hardwire it to 1024. */
#define bits 1024

	if ((U32)keylength > sizeof(self->xkey)) {
		PyErr_SetString(PyExc_ValueError,
				"ARC2 key length must be less than 128 bytes");
		return;
	}

	memcpy(self->xkey, key, keylength);
  
	/* Phase 1: Expand input key to 128 bytes */
	if (keylength < 128) {
		i = 0;
		x = ((U8 *)self->xkey)[keylength-1];
                do {
			x = permute[(x + ((U8 *)self->xkey)[i++]) & 255];
                        ((U8 *)self->xkey)[keylength++] = x;
                } while (keylength < 128);
	}
  
	/* Phase 2 - reduce effective key size to "bits" */
	keylength = (bits+7) >> 3;
	i = 128-keylength;
	x = permute[((U8 *)self->xkey)[i] & (255 >>
					     (7 &
					      ((bits %8 ) ? 8-(bits%8): 0))
		)];
	((U8 *)self->xkey)[i] = x;
  
	while (i--) {
		x = permute[ x ^ ((U8 *)self->xkey)[i+keylength] ];
		((U8 *)self->xkey)[i] = x;
        }
  
	/* Phase 3 - copy to self->xkey in little-endian order */
	i = 63;
	do {
		self->xkey[i] =  ((U8 *)self->xkey)[2*i] +
			(((U8 *)self->xkey)[2*i+1] << 8);
	} while (i--);
}

#undef bits

#include "block_template.c"
